test/e2e/external_gateways.go

```diff
@@ -0,0 +1,482 @@
+		ecmpRetry       int    = 20
+		srcPodName             = "e2e-exgw-src-pod"
+		externalTCPPort        = 80
+		externalUDPPort        = 90
+	)
+
+	f := framework.NewDefaultFramework(svcname)
+
+	var addressesv4, addressesv6 gatewayTestIPs
+
+	ginkgo.BeforeEach(func() {
+		addressesv4, addressesv6 = setupGatewayContainers(f, gwContainer1, gwContainer2, srcPodName, externalUDPPort, externalTCPPort, ecmpRetry)
+
+		// remove the routing external annotation
+		annotateArgs := []string{
+			"annotate",
+			"namespace",
+			f.Namespace.Name,
+			"k8s.ovn.org/routing-external-gws-",
+		}
+		ginkgo.By("Resetting the gw annotation")
+		framework.RunKubectlOrDie(f.Namespace.Name, annotateArgs...)
+	})
+
+	ginkgo.AfterEach(func() {
+		// tear down the containers simulating the gateways
+		deleteClusterExternalContainer(gwContainer1)
+		deleteClusterExternalContainer(gwContainer2)
+	})
+
+	ginkgotable.DescribeTable("Should validate connectivity to multiple external gateways for an ECMP scenario", func(addresses *gatewayTestIPs, icmpToDump string) {
+		if addresses.srcPodIP == "" || addresses.nodeIP == "" {
+			skipper.Skipf("Skipping as pod ip / node ip are not set pod ip %s node ip %s", addresses.srcPodIP, addresses.nodeIP)
+		}
+
+		// annotate the test namespace with multiple gateways defined
+		annotateArgs := []string{
+			"annotate",
+			"namespace",
+			f.Namespace.Name,
+			fmt.Sprintf("k8s.ovn.org/routing-external-gws=%s,%s", addresses.gatewayIPs[0], addresses.gatewayIPs[1]),
+		}
+		framework.Logf("Annotating the external gateway test namespace to container gateways: %s, %s", addresses.gatewayIPs[0], addresses.gatewayIPs[1])
+		framework.RunKubectlOrDie(f.Namespace.Name, annotateArgs...)
+
+		// Verify the gateways and remote loopback addresses are reachable from the pod.
+		// Iterate checking connectivity to the loopbacks on the gateways until tcpdump see
+		// the traffic or 20 attempts fail. Odds of a false negative here is ~ (1/2)^20
+		ginkgo.By("Verifying ecmp connectivity to the external gateways by iterating through the targets")
+
+		// Check for egress traffic to both gateway loopback addresses using tcpdump, since
+		// /proc/net/dev counters only record the ingress interface traffic is received on.
+		// The test will waits until an ICMP packet is matched on the gateways or fail the
+		// test if a packet to the loopback is not received within the timer interval.
+		// If an ICMP packet is never detected, return the error via the specified chanel.
+
+		g := errgroup.Group{}
+		checkPingOnContainer := func(container string) error {
+			_, err := runCommand("docker", "exec", container, "timeout", testTimeout, "tcpdump", "-c", "1", icmpToDump)
+			if err != nil {
+				return err
+			}
+			framework.Logf("ICMP packet successfully detected on gateway %s", container)
+			return nil
+		}
+
+		g.Go(func() error { return checkPingOnContainer(gwContainer1) })
+		g.Go(func() error { return checkPingOnContainer(gwContainer2) })
```

Comment:
is not the pingSync the same, why don't we use the same pattern? the pingSync seems more intuitive

Reply:
That's right. The main reason is how this evolved (and the fact that I realized only lately that the ping goroutines weren't being waited before the test exited).
Will change it to use the same pattern.